### PR TITLE
420 meta jen3

### PR DESCRIPTION
**Pull request by jenlampton**
Another PR for https://github.com/backdrop/backdrop-issues/issues/420#issuecomment-68507351

### CODE REVIEW

File: core/modules/menu/tests/menu.test [hunk 1]
```diff
@@ -508,6 +508,24 @@ class MenuTestCase extends BackdropWebTestCase {
   }
 
   /**
+   * Test administrative users other than user 1 can access the menu parents AJAX callback.
+   */
+  public function testMenuParentsJsAccess() {
+
+    $admin = $this->drupalCreateUser(array('administer menu'));
```
Comment (quicksketch) [inline core/modules/menu/tests/menu.test]: drupal > backdrop

File: core/modules/node/tests/node.test [hunk 2]
```diff
@@ -2493,6 +2493,45 @@ class NodeEntityViewModeAlterTest extends NodeWebTestCase {
...
+  /**
+   * Tests fields that were previously hidden when the view mode is changed.
+   */
+  function testNodeViewModeChangeHiddenField() {
+    // Hide the tags field on the default display
+    $instance = field_info_instance('node', 'field_tags', 'article');
+    $instance['display']['default']['type'] = 'hidden';
+    field_update_instance($instance);
+
+    $web_user = $this->drupalCreateUser(array('create article content', 'edit own article content'));
```
Comment (quicksketch) [inline core/modules/node/tests/node.test]: drupal > backdrop

File: core/modules/node/tests/node.test [hunk 3]
```diff
@@ -2493,6 +2493,45 @@ class NodeEntityViewModeAlterTest extends NodeWebTestCase {
...
+    $this->drupalLogin($web_user);
```
Comment (quicksketch) [inline core/modules/node/tests/node.test]: drupal > backdrop

File: core/modules/entity/entity.module [hunk 4]
```diff
@@ -328,6 +328,57 @@ function entity_prepare_view($entity_type, $entities) {
...
+    $entity_view_mode = $view_mode;
+    if (empty($entity->entity_view_prepared)) {
+
+      // Allow modules to change the view mode.
+      $context = array(
+        'entity_type' => $entity_type,
+        'entity' => $entity,
+        'langcode' => $langcode,
+      );
+      drupal_alter('entity_view_mode', $entity_view_mode, $context);
```
Comment (quicksketch) [inline core/modules/entity/entity.module]: drupal > backdrop

File: core/modules/menu/tests/menu.test [hunk 5]
```diff
@@ -508,6 +508,24 @@ class MenuTestCase extends BackdropWebTestCase {
...
+    $this->drupalLogin($admin);
+    // Just check access to the callback overall, the POST data is irrelevant.
+    $this->drupalGetAJAX('admin/structure/menu/parents');
```
Comment (quicksketch) [inline core/modules/menu/tests/menu.test]: drupal > backdrop

File: core/modules/menu/tests/menu.test [hunk 6]
```diff
@@ -508,6 +508,24 @@ class MenuTestCase extends BackdropWebTestCase {
   }
 
   /**
+   * Test administrative users other than user 1 can access the menu parents AJAX callback.
+   */
+  public function testMenuParentsJsAccess() {
+
+    $admin = $this->drupalCreateUser(array('administer menu'));
+    $this->drupalLogin($admin);
```
Comment (quicksketch) [inline core/modules/menu/tests/menu.test]: drupal > backdrop

File: core/modules/menu/tests/menu.test [hunk 7]
```diff
@@ -508,6 +508,24 @@ class MenuTestCase extends BackdropWebTestCase {
...
+    // Just check access to the callback overall, the POST data is irrelevant.
+    $this->drupalGetAJAX('admin/structure/menu/parents');
+    $this->assertResponse(200);
+
+    // Do standard user tests.
+    // Login the user.
+    $this->drupalLogin($this->std_user);
```
Comment (quicksketch) [inline core/modules/menu/tests/menu.test]: drupal > backdrop

File: core/modules/menu/tests/menu.test [hunk 8]
```diff
@@ -508,6 +508,24 @@ class MenuTestCase extends BackdropWebTestCase {
...
+    $this->drupalGetAJAX('admin/structure/menu/parents');
```
Comment (quicksketch) [inline core/modules/menu/tests/menu.test]: drupal > backdrop

File: core/modules/node/tests/node.test [hunk 9]
```diff
@@ -2493,6 +2493,45 @@ class NodeEntityViewModeAlterTest extends NodeWebTestCase {
...
+
+    // Create a node.
+    $edit = array();
+    $langcode = LANGUAGE_NONE;
+    $edit["title"] = $this->randomName(8);
+    $edit["body[$langcode][0][value]"] = 'Data that should appear only in the body for the node.';
+    $edit["body[$langcode][0][summary]"] = 'Extra data that should appear only in the teaser for the node.';
+    $edit["field_tags[$langcode]"] = 'Extra tag';
+    $this->drupalPost('node/add/article', $edit, t('Save'));
```
Comment (quicksketch) [inline core/modules/node/tests/node.test]: drupal > backdrop

File: core/modules/node/tests/node.test [hunk 10]
```diff
@@ -2493,6 +2493,45 @@ class NodeEntityViewModeAlterTest extends NodeWebTestCase {
...
+
+    $node = $this->drupalGetNodeByTitle($edit["title"]);
```
Comment (quicksketch) [inline core/modules/node/tests/node.test]: drupal > backdrop

File: core/modules/node/tests/node.test [hunk 11]
```diff
@@ -2493,6 +2493,45 @@ class NodeEntityViewModeAlterTest extends NodeWebTestCase {
...
+
+    // Set the flag to alter the view mode and view the node.
+    variable_set('node_test_change_view_mode', 'teaser');
+    $this->drupalGet('node/' . $node->nid);
```
Comment (quicksketch) [inline core/modules/node/tests/node.test]: drupal > backdrop

**Comment by jenlampton** [comment]
cleaned up all the drupal -> backdrops. Pushed up again, testing now.

File: core/includes/theme.inc [hunk 12]
```diff
@@ -1181,7 +1184,12 @@ function theme($hook, $variables = array()) {
     if (isset($info['path'])) {
       $template_file = $info['path'] . '/' . $template_file;
     }
-    $output = $render_function($template_file, $variables);
+    if (variable_get('theme_debug', FALSE)) {
```
Comment (quicksketch) [inline core/includes/theme.inc]: This variable needs to be converted to config in system.core.json.

File: core/modules/block/block.install [hunk 13]
```diff
@@ -127,6 +127,22 @@ function block_update_1002() {
 }
 
 /**
+ * Increase {block}.title length to 255 characters.
```
Comment (quicksketch) [inline core/modules/block/block.install]: This should be {block_custom}.

File: core/modules/simpletest/tests/theme.test [hunk 14]
```diff
@@ -649,3 +739,68 @@ class ThemeDatetime extends BackdropWebTestCase {
...
+    theme_enable(array('test_theme'));
+  }
+
+  /**
+   * Tests debug markup added to template output.
+   */
+  function testDebugOutput() {
+    variable_set('theme_default', 'test_theme');
+    // Enable the debug output.
+    variable_set('theme_debug', TRUE);
```
Comment (quicksketch) [inline core/modules/simpletest/tests/theme.test]: variable -> config

File: core/includes/theme.inc (hunk 12 above)
Comment (jenlampton) [inline core/includes/theme.inc]: I don't think it's config, I think it's state.  Updating...

File: core/modules/block/block.install (hunk 13 above)
Comment (jenlampton) [inline core/modules/block/block.install]: fixed.

File: core/modules/simpletest/tests/theme.test (hunk 14 above)
Comment (jenlampton) [inline core/modules/simpletest/tests/theme.test]: state?

File: core/includes/theme.inc [hunk 15]
```diff
@@ -1181,7 +1184,12 @@ function theme($hook, $variables = array()) {
     if (isset($info['path'])) {
       $template_file = $info['path'] . '/' . $template_file;
     }
-    $output = $render_function($template_file, $variables);
+    if (state_get('theme_debug', FALSE)) {
+      $output = _theme_render_template_debug($render_function, $template_file, $variables, $extension);
+    }
+    else {
+      $output = $render_function($template_file, $variables);
+    }
```
Comment (quicksketch) [inline core/includes/theme.inc:1192]: Theme debug still needs to be converted to use config. We should save it in system.core.json.

Comment (jenlampton) [inline core/includes/theme.inc:1192]: think i got it this time...

File: core/modules/simpletest/tests/theme.test [hunk 16]
```diff
@@ -533,6 +533,96 @@ class ThemeHeadTag extends BackdropUnitTestCase {
...
+  }
+
+  /**
+   * Tests system #type 'html_tag'.
+   */
+  function testHtmlTag() {
+    $elements = array(
+      // Test auto-closure meta tag generation.
+      array(
+        'name' => "#type 'html_tag' auto-closure meta tag generation",
```
Comment (quicksketch) [inline core/modules/simpletest/tests/theme.test]: This test needs to be converted to be "head_tag". They're currently failing. :\

File: core/modules/simpletest/tests/theme.test [hunk 17]
```diff
@@ -649,3 +739,68 @@ class ThemeDatetime extends BackdropWebTestCase {
...
+ */
+class ThemeDebugMarkupTestCase extends BackdropWebTestCase {
+
+  public static function getInfo() {
+    return array(
+      'name' => 'Theme debug markup',
+      'description' => 'Tests theme debug markup output.',
+      'group' => 'Theme',
+    );
+  }
```
Comment (quicksketch) [inline core/modules/simpletest/tests/theme.test]: This needs to be converted to a *.tests.info file.

**Comment by jenlampton** [comment]
pushed a new version

**Pull request merged**